### ISSUE TITLE
Maxixum query size or binds size?

### ISSUE THREAD

**Issue by greglearns** [sfackler/rust-postgres#356]
When I write an INSERT query that is 289,151 characters long, with 40,000 bind variables, I get this error, but can't figure out where it is coming from (I've looked through the code):

`type conversion error: value too large to transmit`

Each row inserted is something like 5KB characters big as JSON. If I do the math right, then:
40,000 rows * 5 KB of characters * 4 bytes in a 32-bit character = 800,000,000 bytes = 800 MB (is that correct?)

It's a big INSERT query, but I can't see in the code where I am hitting any limits that would trigger the error.

**Comment by sfackler** [comment]
The Postgres communication protocol uses a signed 16 bit integer to identify the number of query parameters, so you can't execute a query that contains more than 32,768 of them. A COPY query is generally a much better way to insert large amounts of data.

**Comment by greglearns** [comment]
How does one safely escape COPY queries if you are adding user-submitted data?

**Comment by sfackler** [comment]
You could use the csv format and the [csv](https://crates.io/crates/csv) crate, or the binary format and the [postgres-binary-copy](https://crates.io/crates/postgres-binary-copy) crate.

**Comment by greglearns** [comment]
I didn't know about the postgres-binary-copy crate -- nice! So, it is safe to pass any binary data into that crate?

(and, thank you so much for all of your postgres work)

**Comment by sfackler** [comment]
Yep, using that crate for a binary format copy is safe in the same way that normal query parameters are.

**Comment by greglearns** [comment]
Wow. That is nice. To date, I've been building (very long) strings for the query params, so using the binary crate sounds much much easier. Is there any limits on the amount of data one can pass in?

**Comment by sfackler** [comment]
I don't think so. COPY queries are kind of "special" in that they switch the communication protocol into a streaming mode specifically designed to support very large writes.

**Comment by greglearns** [comment]
thank you
